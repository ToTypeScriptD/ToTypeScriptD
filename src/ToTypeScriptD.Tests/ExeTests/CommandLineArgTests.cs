using ApprovalTests;
using System.IO;
using Xunit;

namespace ToTypeScriptD.Tests.ExeTests
{
    public class CommandLineArgTests : ExeTestBase
    {
        [Fact]
        public void ExeShouldGenerateHelpOnEmptyInput()
        {
            var result = Execute("").StdOut.StripVersionFromOutput();
            Approvals.Verify(result);
        }

        [Fact]
        public void ExeShouldGenerateHelpWithHelpArgs()
        {
            var result = Execute("--help").StdOut.StripVersionFromOutput();
            Approvals.Verify(result);
        }

        [Fact]
        public void ExeShouldGenerateHelpForDotNet()
        {
            var result = Execute(Options.DotNetCommandName).StdOut.StripVersionFromOutput();
            Approvals.Verify(result);
        }

        [Fact]
        public void ExeShouldGenerateHelpForWinMD()
        {
            var result = Execute(Options.WinmdCommandName).StdOut.StripVersionFromOutput();
            Approvals.Verify(result);
        }

        [Fact]
        public void ExeShouldGenerateHelpForDotNetWithBadInput()
        {
            var result = Execute(Options.DotNetCommandName + " --nogohere").StdOut.StripVersionFromOutput();
            Approvals.Verify(result);
        }

        [Fact]
        public void ExeShouldGenerateHelpForWinMDWithBadInput()
        {
            var result = Execute(Options.WinmdCommandName + " --nogohere").StdOut.StripVersionFromOutput();
            Approvals.Verify(result);
        }

        [Fact]
        public void ExeShouldGenerateOutputForMultipleWinmdFiles()
        {
            var result = Execute(@"winmd C:\Windows\System32\WinMetadata\Windows.Foundation.winmd C:\Windows\System32\WinMetadata\Windows.Networking.winmd");

            (result.StdOut.Length > 100).ShouldBeTrue(result.StdOut.Length + " should be greater than 100");
        }

        [Fact]
        public void VerifyExeGeneratedOutputForMultipleWinmdFiles()
        {
            var result = Execute(@"winmd C:\Windows\System32\WinMetadata\Windows.Foundation.winmd C:\Windows\System32\WinMetadata\Windows.System.winmd");

            result.StdOut.Verify();
        }

        [Fact]
        [ApprovalTests.Reporters.UseReporter(typeof(ApprovalTests.Reporters.P4MergeReporter))]
        public void ExeDuplicateAssemblyShouldStillOnlyGenerateOne()
        {
            var resultDup = Execute(@"winmd C:\Windows\System32\WinMetadata\Windows.Foundation.winmd C:\Windows\System32\WinMetadata\Windows.Foundation.winmd");
            var resultNonDup = Execute(@"winmd C:\Windows\System32\WinMetadata\Windows.Foundation.winmd");

            resultNonDup.ToString().StripHeaderGarbageromOutput()
                .DiffWith(resultDup.ToString().StripHeaderGarbageromOutput());
        }

        [Fact]
        public void ExeShouldBeAbleToGenerateSpecialTypes()
        {
            var resultDup = Execute("winmd --specialTypes");

            resultDup.Verify();
        }


        [Fact]
        public void ExeShouldGenerateWith2SpaceIndentation()
        {
            var resultDup = Execute("winmd --specialTypes ToTypeScriptD.Native.winmd --indentWith SpaceX2");

            resultDup.Verify();
        }

        [Fact]
        public void ExeShouldGiveHelpfulErrorWhenFilesNotFoundInUnknownDirectory()
        {
            var resultDup = Execute(@"winmd C:\TypeScriptD\TypeScriptD\TypeScriptD\Foo.dll C:\TypeScriptD\TypeScriptD\TypeScriptD\Foo.dll");

            resultDup.Verify();
        }

        [Fact]
        public void ExeShouldGiveHelpfulErrorWhenFilesNotFound()
        {
            var resultDup = Execute(@"winmd C:\TypeScriptD_FileNotFound_ThisShouldNotExistOnYourSystem.dll");

            resultDup.Verify();
        }

        [Fact]
        public void ExeShouldApplyRegexFilterOnTypes()
        {
            var resultDup = Execute(@"DotNet ..\..\..\bin\ToTypeScriptD.TestAssembly.CSharp.dll --regexFilter 'ToTypeScriptD.TestAssembly.CSharp.NamespaceSample|CSharp.NamespaceSample'");

            resultDup.Verify();
        }

    }
}
