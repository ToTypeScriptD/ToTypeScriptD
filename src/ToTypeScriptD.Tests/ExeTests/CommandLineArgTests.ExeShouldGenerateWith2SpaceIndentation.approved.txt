//****************************************************************
//  Generated by:  ToTypeScriptD
//  Website:       http://github.com/ToTypeScriptD/ToTypeScriptD
//  Version:       v0.0.0000.0000 SHA1:0000000 - Debug
//  Date:          mm/dd/YYYY H:MM:SS PM
//
//  Assemblies:
//    ToTypeScriptD.Native.winmd
//
//****************************************************************




/*
 * This TypeScript definition file contains some core WinRT type 
 * definitions that the TypeScriptD output may need for things like async etc.
 *
 * Include this definition with the rest of your *.d.ts includes.
 *
 */


declare module ToTypeScriptD.WinRT {
  export interface IPromise<TResult> {
    then<U>(success?: (value: TResult) => IPromise<U>, error?: (error: any) => IPromise<U>, progress?: (progress: any) => void): IPromise<U>;
    then<U>(success?: (value: TResult) => IPromise<U>, error?: (error: any) => U, progress?: (progress: any) => void): IPromise<U>;
    then<U>(success?: (value: TResult) => U, error?: (error: any) => IPromise<U>, progress?: (progress: any) => void): IPromise<U>;
    then<U>(success?: (value: TResult) => U, error?: (error: any) => U, progress?: (progress: any) => void): IPromise<U>;
    done<U>(success?: (value: TResult) => any, error?: (error: any) => any, progress?: (progress: any) => void): void;
    cancel(): void;

    //operation: Windows.Foundation.IAsyncOperation<T>;
  }
}



declare module ToTypeScriptD.Native {

  export class AsyncClass {
    constructor();
    getStringAsync(value: string): Windows.Foundation.IAsyncOperation<string>;
  }

  export class ClassWithEventHandler {
    constructor();
    doSomething(): void;
    doSomethingTyped(): void;
    // Events
    addEventListener(eventName: string, listener: any): void;
    removeEventListener(eventName: string, listener: any): void;
    addEventListener(eventName: "somethinghappened", listener: ToTypeScriptD.Native.SomethingHappenedEventHandler): void;
    removeEventListener(eventName: "somethinghappened", listener: ToTypeScriptD.Native.SomethingHappenedEventHandler): void;
    onsomethinghappened: (ev: ToTypeScriptD.Native.SomethingHappenedEventHandler) => void;
    addEventListener(eventName: "sampletyped", listener: Windows.Foundation.TypedEventHandler<ToTypeScriptD.Native.ClassWithEventHandler,ToTypeScriptD.Native.SampleEnum>): void;
    removeEventListener(eventName: "sampletyped", listener: Windows.Foundation.TypedEventHandler<ToTypeScriptD.Native.ClassWithEventHandler,ToTypeScriptD.Native.SampleEnum>): void;
    onsampletyped: (ev: Windows.Foundation.TypedEventHandler<ToTypeScriptD.Native.ClassWithEventHandler,ToTypeScriptD.Native.SampleEnum>) => void;

  }

  export class CXReturnTypes {
    constructor();
    getString(): string;
    getInt32(): number;
    getUInt32(): number;
    getGuid(): string;
    getByte(): number;
    getBytes(): any;
    getBoolean(): boolean;
    methodWithSomeOutParameters(in1: number, in2: string): CXReturnTypes_methodWithSomeOutParameters_OUT;
    methodWithSomeOutParametersButVoidResult(in1: number, in2: string): CXReturnTypes_methodWithSomeOutParametersButVoidResult_OUT;
    overloadedMethodWithNumberTypeParams(value: number): string;
    getIterableOfInts(): Windows.Foundation.Collections.IIterable<number>;
  }

  interface CXReturnTypes_methodWithSomeOutParameters_OUT {
    __returnValue: number;
    out1: number;
    out2: string;
    out3: number[];
  }

  interface CXReturnTypes_methodWithSomeOutParametersButVoidResult_OUT {
    out1: number;
    out2: string;
    out3: number[];
  }

  enum SampleEnum {
    a,
    b,
    c,
    d
  }

  export class SampleEnumClass {
    constructor();
    methowWithEnumParameter(value: ToTypeScriptD.Native.SampleEnum): ToTypeScriptD.Native.SampleEnum;
  }

  enum SampleEnumNumbered {
    a,
    b,
    d,
    c
  }

  export interface SomethingHappenedEventHandler {
    target: ToTypeScriptD.Native.ClassWithEventHandler;
    detail: any[];
    type: string;
  }

}

